prometheus-metrics-model/src/main/java/io/prometheus/metrics/model/snapshots/PrometheusNaming.java

```diff
@@ -280,4 +371,368 @@ private static String replaceIllegalCharsInUnitName(String name) {
     }
     return new String(sanitized);
   }
+
+  /**
+   * Escapes the given metric names and labels with the given
+   * escaping scheme.
+   */
+  public static MetricSnapshot escapeMetricSnapshot(MetricSnapshot v, EscapingScheme scheme) {
+    if (v == null) {
+      return null;
+    }
+
+    if (scheme == EscapingScheme.NO_ESCAPING) {
+      return v;
+    }
+
+    String outName;
+
+    // If the name is null, copy as-is, don't try to escape.
+    if (v.getMetadata().getPrometheusName() == null || isValidLegacyMetricName(v.getMetadata().getPrometheusName())) {
+      outName = v.getMetadata().getPrometheusName();
+    } else {
+      outName = escapeName(v.getMetadata().getPrometheusName(), scheme);
+    }
+
+    List<DataPointSnapshot> outDataPoints = new ArrayList<>();
+
+    for (DataPointSnapshot d : v.getDataPoints()) {
+      if (!metricNeedsEscaping(d)) {
+        outDataPoints.add(d);
+        continue;
+      }
+
+      Labels.Builder outLabelsBuilder = Labels.builder();
+
+      for (Label l : d.getLabels()) {
+        if (METRIC_NAME_LABEL.equals(l.getName())) {
+          if (l.getValue() == null || isValidLegacyMetricName(l.getValue())) {
+            outLabelsBuilder.label(l.getName(), l.getValue());
+            continue;
+          }
+          outLabelsBuilder.label(l.getName(), escapeName(l.getValue(), scheme));
+          continue;
+        }
+        if (l.getName() == null || isValidLegacyMetricName(l.getName())) {
+          outLabelsBuilder.label(l.getName(), l.getValue());
+          continue;
+        }
+        outLabelsBuilder.label(escapeName(l.getName(), scheme), l.getValue());
+      }
+
+      Labels outLabels = outLabelsBuilder.build();
+      DataPointSnapshot outDataPointSnapshot = null;
```

Comment:
better extract a method that returns the snapshot rather than reassigning the variable